File: agent-service/src/agents/testing/testing-agent.ts
/**
 * TestingAgent — runs existing tests and optionally generates new ones via LLM.
 *
 * Lifecycle:
 *  onInit    — build LLM, detect test framework in each workspace
 *  onExecute — run tests; if tests fail or none exist, generate new tests via LLM
 *  onCleanup — no-op
 *
 * Events published via AgentEventBus:
 *  "tests.started"  { agentId, executionId }
 *  "tests.passed"   { agentId, executionId, passed, failed, durationMs }
 *  "tests.failed"   { agentId, executionId, passed, failed, durationMs }
 *  "tests.generated" { agentId, executionId, filesGenerated }
 */
import * as fs from "fs/promises";
import * as path from "path";
import { BaseAgent } from "../../core/agent-framework/base-agent";
import type { AgentContext } from "../../core/agent-framework/context";
import type { AgentArtifact } from "../coordinator/drivers/base";
import { getAgentEventBus } from "../../core/agent-framework/event-bus";
import { buildChatModel } from "../../core/llm";
import { detectFramework, runTests } from "./test-runner";
import type { TestRunResult } from "./test-runner";
import type { BaseChatModel } from "@langchain/core/language_models/chat_models";
import { HumanMessage, SystemMessage } from "@langchain/core/messages";

export interface TestingAgentConfig {
  llmProvider: string;
  llmModel: string;
  llmApiKey: string;
  llmBaseUrl?: string;
  /** If true, generate tests when none are found. Default: true. */
  generateTests?: boolean;
}

const TEST_GEN_SYSTEM_PROMPT = `You are an expert software engineer specializing in test-driven development.

Given a task description and source code files, generate comprehensive tests that:
1. Cover happy paths, edge cases, and error conditions
2. Follow the detected test framework conventions
3. Are self-contained and runnable without mocks unless necessary

Respond with a JSON array of files to create:
[
  {
    "path": "tests/unit/example.test.ts",
    "content": "// full file content here"
  }
]

Respond with ONLY valid JSON. No markdown fences.`;

export class TestingAgent extends BaseAgent {
  private llm!: BaseChatModel;

  constructor(
    id: string,
    private readonly config: TestingAgentConfig,
  ) {
    super(id, "tester");
  }

  protected async onInit(_ctx: AgentContext): Promise<void> {
    this.llm = buildChatModel({
      provider: this.config.llmProvider,
      model: this.config.llmModel,
      apiKey: this.config.llmApiKey,
      baseUrl: this.config.llmBaseUrl,
    });
  }

  protected async onExecute(ctx: AgentContext): Promise<void> {
    const bus = getAgentEventBus();
    await bus.publish("tests.started", { agentId: this.id, executionId: ctx.executionId });

    const workspaces = ctx.workspaces ?? {};
    const allResults: TestRunResult[] = [];
    const generatedFiles: string[] = [];

    for (const [repoName, localPath] of Object.entries(workspaces)) {
      const framework = await detectFramework(localPath);

      if (!framework) {
        // No known test framework — optionally generate tests
        if (this.config.generateTests !== false) {
          const generated = await this.generateTests(ctx, localPath, "jest");
          generatedFiles.push(...generated.map((f) => `${repoName}/${f}`));
        }
        continue;
      }

      const result = await runTests(localPath, framework);
      allResults.push(result);

      if (!result.success && this.config.generateTests !== false) {
        // Tests failed — generate additional tests to improve coverage
        const generated = await this.generateTests(ctx, localPath, framework.name);
        generatedFiles.push(...generated.map((f) => `${repoName}/${f}`));
      }
    }

    const totalPassed = allResults.reduce((s, r) => s + r.passed, 0);
    const totalFailed = allResults.reduce((s, r) => s + r.failed, 0);
    const totalDuration = allResults.reduce((s, r) => s + r.durationMs, 0);
    const overallSuccess = allResults.every((r) => r.success) && allResults.length > 0;

    const summary = {
      passed: totalPassed,
      failed: totalFailed,
      durationMs: totalDuration,
      results: allResults,
      generatedFiles,
    };

    this._result = {
      success: overallSuccess || generatedFiles.length > 0,
      output: JSON.stringify(summary),
      artifacts: [
        {
          type: "test_result",
          title: `Test Results — ${totalPassed} passed, ${totalFailed} failed`,
          content: JSON.stringify(summary),
          content_type: "application/json",
          metadata: {
            passed: totalPassed,
            failed: totalFailed,
            skipped: 0,
            total: totalPassed + totalFailed,
            duration_ms: totalDuration,
          },
        } satisfies AgentArtifact,
      ],
    };

    const eventTopic = overallSuccess ? "tests.passed" : "tests.failed";
    await bus.publish(eventTopic, {
      agentId: this.id,
      executionId: ctx.executionId,
      passed: totalPassed,
      failed: totalFailed,
      durationMs: totalDuration,
    });

    if (generatedFiles.length > 0) {
      await bus.publish("tests.generated", {
        agentId: this.id,
        executionId: ctx.executionId,
        filesGenerated: generatedFiles.length,
      });
    }
  }

  protected async onCleanup(): Promise<void> {
    // Nothing to release
  }

  private async generateTests(
    ctx: AgentContext,
    workspacePath: string,
    frameworkHint: string,
  ): Promise<string[]> {
    const taskDesc = ctx.task?.description ?? "Generate tests for the codebase.";

    // Collect a sample of source files for context
    const srcFiles = await this.collectSourceFiles(workspacePath, 5);
    const srcContext = srcFiles.map(([filePath, content]) =>
      `### ${filePath}\n\`\`\`\n${content.slice(0, 2000)}\n\`\`\``
    ).join("\n\n");

    const response = await this.llm.invoke([
      new SystemMessage(TEST_GEN_SYSTEM_PROMPT),
      new HumanMessage(
        `Task: ${taskDesc}\n\nTest framework: ${frameworkHint}\n\nSource files:\n${srcContext}`
      ),
    ]);

    const content = typeof response.content === "string"
      ? response.content
      : JSON.stringify(response.content);

    try {
      const jsonMatch = content.match(/\[[\s\S]*\]/);
      const files = JSON.parse(jsonMatch?.[0] ?? content) as Array<{ path: string; content: string }>;

      const written: string[] = [];
      for (const file of files) {
        const absPath = path.join(workspacePath, file.path);
        await fs.mkdir(path.dirname(absPath), { recursive: true });
        await fs.writeFile(absPath, file.content, "utf-8");
        written.push(file.path);
      }
      return written;
    } catch {
      return [];
    }
  }

  private async collectSourceFiles(
    dir: string,
    maxFiles: number,
  ): Promise<[string, string][]> {
    const results: [string, string][] = [];
    const IGNORE_DIRS = new Set(["node_modules", ".git", "dist", "build", ".next", "coverage"]);
    const SRC_EXTS = new Set([".ts", ".tsx", ".js", ".jsx", ".py", ".go", ".rs"]);

    async function walk(current: string): Promise<void> {
      if (results.length >= maxFiles) return;
      const entries = await fs.readdir(current, { withFileTypes: true }).catch(() => null);
      if (!entries) return;

      for (const entry of entries) {
        if (results.length >= maxFiles) return;
        if (entry.isDirectory()) {
          if (!IGNORE_DIRS.has(String(entry.name))) await walk(path.join(current, String(entry.name)));
        } else if (SRC_EXTS.has(path.extname(String(entry.name)))) {
          const fullPath = path.join(current, String(entry.name));
          try {
            const content = await fs.readFile(fullPath, "utf-8");
            results.push([path.relative(dir, fullPath), content]);
          } catch { /* skip unreadable files */ }
        }
      }
    }

    await walk(dir);
    return results;
  }
}

File: agent-service/src/agent_service/agents/testing/testing_agent.py
from __future__ import annotations

import json
from typing import List, Optional

from pydantic import BaseModel

from agent_service.core.agent_framework.base_agent import BaseAgent, AgentResult as BaseAgentResult
from agent_service.core.agent_framework.context import AgentContext
from agent_service.core.agent_framework.event_bus import get_agent_event_bus
from agent_service.core.llm import build_chat_model
from agent_service.agents.testing.tools import build_testing_tools
from langchain_core.messages import HumanMessage, ToolMessage
from langgraph.prebuilt import create_react_agent


class TestingAgentConfig(BaseModel):
    llmProvider: str = "openai"
    llmModel: str = "gpt-4o"
    llmApiKey: str = ""
    llmBaseUrl: Optional[str] = None
    llmTemperature: Optional[float] = None
    llmMaxTokens: Optional[int] = None
    llmTopP: Optional[float] = None
    llmFrequencyPenalty: Optional[float] = None
    llmPresencePenalty: Optional[float] = None
    systemPrompt: Optional[str] = None
    systemPromptMode: str = "override"  # "override" | "extend"
    mcpServers: List[dict] = []
    skills: List[dict] = []
    subAgentIds: List[str] = []
    generateTests: bool = True


def _compose_prompt(builtin: str, custom: Optional[str], mode: str) -> str:
    """Return the effective system prompt based on mode and user-supplied prompt."""
    if not custom:
        return builtin
    if mode == "override":
        return custom
    return f"{builtin}\n\n{custom}"


TEST_SYSTEM_PROMPT = """\
You are an expert software engineer specialising in test-driven development and quality assurance.

Your workflow:
1. **Detect the test framework** — read `package.json`, `pyproject.toml`, `go.mod`, or similar to identify the language and test runner (jest, pytest, go test, etc.)
2. **Inspect the source code** — use `read_file` to understand what has been implemented
3. **Write tests** — use `write_file` to create comprehensive test files covering:
   - Happy paths
   - Edge cases and boundary conditions
   - Error conditions and failure modes
4. **Run the tests** — use `run_shell` to execute the test suite and capture results
5. **Iterate if needed** — if tests fail due to test file issues (not implementation bugs), fix them and re-run
6. **Finish** — call `finish` with the final results

When calling `finish`:
- `passed`: true only if the test suite exits with a 0 exit code
- `summary`: brief description of what was tested and the outcome
- `tests_run`: total number of test cases executed
- `failures`: list of failing test names or error messages (empty if all passed)"""


class TestingAgent(BaseAgent):
    def __init__(self, id: str, config: TestingAgentConfig) -> None:
        super().__init__(id, "tester")
        self._config = config
        self._llm = None

    async def on_init(self, ctx: AgentContext) -> None:
        self._llm = build_chat_model(
            provider=self._config.llmProvider,
            model=self._config.llmModel,
            api_key=self._config.llmApiKey,
            base_url=self._config.llmBaseUrl,
            temperature=self._config.llmTemperature,
            max_tokens=self._config.llmMaxTokens,
            top_p=self._config.llmTopP,
            frequency_penalty=self._config.llmFrequencyPenalty,
            presence_penalty=self._config.llmPresencePenalty,
        )

    async def on_execute(self, ctx: AgentContext) -> None:
        bus = get_agent_event_bus()
        await bus.publish("tests.started", {"agentId": self.id, "executionId": ctx.executionId})

        task_desc = ctx.task.description if ctx.task else "Test the implemented code."
        task_title = ctx.task.title if ctx.task else "Testing"

        effective_system = _compose_prompt(
            TEST_SYSTEM_PROMPT, self._config.systemPrompt, self._config.systemPromptMode
        )

        tools = build_testing_tools(ctx.workspaces or {})
        graph = create_react_agent(self._llm, tools, prompt=effective_system)

        human_content = (
            f"Task: {task_title}\n{task_desc}\n\n"
            f"Please detect the test framework, write comprehensive tests for the implemented code, "
            f"run them, and call finish() with the results."
        )

        result = await graph.ainvoke(
            {"messages": [HumanMessage(content=human_content)]},
            {"recursion_limit": 60},
        )

        summary = _extract_finish_result(result.get("messages", []))

        overall_success = summary.get("passed", False)
        tests_run = summary.get("tests_run", 0)
        failures = summary.get("failures", [])

        self._result = BaseAgentResult(
            success=overall_success,
            output=json.dumps(summary),
            artifacts=[{
                "type": "test_result",
                "title": (
                    f"Test Results — {tests_run - len(failures)} passed, {len(failures)} failed"
                    if tests_run
                    else ("Tests Passed" if overall_success else "Tests Failed")
                ),
                "content": json.dumps(summary),
                "content_type": "application/json",
                "metadata": {
                    "passed": tests_run - len(failures),
                    "failed": len(failures),
                    "skipped": 0,
                    "total": tests_run,
                    "duration_ms": 0,
                },
            }],
        )

        event_topic = "tests.passed" if overall_success else "tests.failed"
        await bus.publish(event_topic, {
            "agentId": self.id,
            "executionId": ctx.executionId,
            "testsRun": tests_run,
            "failures": len(failures),
        })

    async def on_cleanup(self) -> None:
        pass


def _extract_finish_result(messages: list) -> dict:
    """
    Find the last ToolMessage from the 'finish' tool and parse its JSON.
    Falls back to a safe default if not found.
    """
    for msg in reversed(messages):
        if isinstance(msg, ToolMessage) and getattr(msg, "name", None) == "finish":
            try:
                data = json.loads(msg.content)
                if isinstance(data, dict) and "passed" in data:
                    return data
            except (json.JSONDecodeError, TypeError):
                pass

    return {
        "passed": False,
        "summary": "Testing agent did not call finish(). Manual inspection required.",
        "tests_run": 0,
        "failures": [],
    }
